### PR TITLE
[MLIR][Transforms] Fix dialect conversion inverse mapping

### PR DESCRIPTION
**Pull request by zyx-billy**
Inverse mapping needs to be updated for the result that was remapped (it was previously only updated halfway).

### CODE REVIEW

File: mlir/lib/Transforms/Utils/DialectConversion.cpp [hunk 1]
```diff
@@ -2527,6 +2527,7 @@ LogicalResult OperationConverter::legalizeConvertedOpResultTypes(
           opReplacement->getConverter());
       rewriterImpl.mapping.map(result, castValue);
       inverseMapping[castValue].push_back(result);
+      llvm::erase(inverseMapping[newValue], result);
```
Comment (zyx-billy) [inline mlir/lib/Transforms/Utils/DialectConversion.cpp:2530]: This is probably a very short list. Open to changing it to a Set though.

**Comment by zyx-billy** [comment]
@raikonenfnu I saw that you were seeing downstream issues here too. Do you want to see if this fixes it too? I'm just working on a test case right now.

**Comment by raikonenfnu** [comment]
Will try this out, thanks! @zyx-billy  :)

File: mlir/test/Transforms/test-legalize-target-materialization-no-uses.mlir [hunk 2]
```diff
@@ -25,3 +25,10 @@ func.func @foo() {
   "test.type_consumer"(%2) : (i64) -> ()
   return
 }
+
```
Comment (matthias-springer) [inline mlir/test/Transforms/test-legalize-target-materialization-no-uses.mlir:28]: Can you add `// -----` here and `-split-input-file` in the `// RUN:` line?

File: mlir/lib/Transforms/Utils/DialectConversion.cpp (hunk 1 above)
Comment (matthias-springer) [reply mlir/lib/Transforms/Utils/DialectConversion.cpp:2530]: I agree, list is usually very short, so a vector should be fine.

**Review by matthias-springer — APPROVED** [review]
Good catch! This is the right fix!

`rewriterImpl.mapping.map(result, castValue)` overwrites the existing mapping of `result`. So the old mapping must be removed from `inverseMapping`.

**Pull request merged**